### ISSUE TITLE
Updated kv_cache?

### ISSUE THREAD

**Issue by toanbh10032** [ambisinister/mla-experiments#2]
in GPT.py / class GPTModel(torch.nn.Module): you return logits, updated_kv_cache
```
 updated_kv_cache = []
        for i, layer in enumerate(self.layers):
            if kv_cache is not None:
                layer_cache = kv_cache[i]
            else:
                layer_cache = None
            x, new_kv = layer(x, kv_cache=layer_cache, past_length=past_length)
            updated_kv_cache.append(new_kv)

        logits = self.fc_out(x)

        return logits, updated_kv_cache
```

but in training.py: 
```
 # https://pytorch.org/docs/stable/amp.html
        with torch.autocast(device_type="cuda", dtype=torch.float16):
            out, _ = model(dat)
            out = out.permute(0, 2, 1)
            loss = loss_fn(out, targ)
            loss = loss / acc_steps
```

The updated_kv_cache will be reset to empty whenever we call GPPTModel function?
the kv_cache seems not to be reused, so I don't see why we need the updated_kv_cache step.  And, I don't see how we cache the kv (kv cache = none at every step)
```
def forward(self, x, kv_cache=None, past_length=0):
        normed = self.norm1(x)
        if kv_cache is not None:
            mh_x, kv = self.mha(normed, kv_cache=kv_cache, past_length=past_length)
        else:
            mh_x, kv = self.mha(normed)
        x = x + mh_x
        x = x + self.dropout(self.fc2(self.act(self.fc1(self.norm2(x)))))
        return x, kv
```
Can you explain if something is wrong or if I misunderstood somewhere?

**Comment by WessonZZ** [comment]
In every inference step, we can pass the last step's KV cache into the "kv_cache" key words in "forward(self, x, kv_cache=None, past_length=0)", then, we do not need to recompute previous steps' KV, and we only need to concat this steps' KV into the "kv_cache". In general, we will always reuse the KV cache in inference stage.
